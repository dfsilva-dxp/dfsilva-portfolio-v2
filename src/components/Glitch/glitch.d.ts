export interface IGlitchProps {
  columns?: "one" | "two" | "three";
}
