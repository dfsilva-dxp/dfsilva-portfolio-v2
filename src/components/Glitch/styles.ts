import styled, { css, DefaultTheme } from "styled-components";

import { IGlitchProps } from "./glitch";

const gridModifiers = {
  one: () => css`
    grid-template-columns: 1fr;
  `,
  two: (theme: DefaultTheme) => css`
    grid-template-columns: repeat(2, 1fr);

    &::before {
      content: "";
      border-right: 1px solid ${theme.colors.glitch};
    }
  `,
  three: (theme: DefaultTheme) => css`
    grid-template-columns: repeat(3, 1fr);

    &::before,
    &::after {
      content: "";
      border-right: 1px solid ${theme.colors.glitch};
    }
  `
};

export const Grid = styled.div<IGlitchProps>`
  ${({ theme, columns }) => css`
    position: absolute;
    height: 100%;
    border-left: 1px solid ${theme.colors.glitch};
    border-right: 1px solid ${theme.colors.glitch};
    left: ${theme.spacings.xsmall};
    right: ${theme.spacings.xsmall};
    z-index: ${theme.layers.glitch};
    display: grid;

    ${!!columns && gridModifiers[columns](theme)}
  `}
`;
