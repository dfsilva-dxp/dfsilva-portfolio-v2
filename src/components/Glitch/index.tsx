import { IGlitchProps } from "./glitch";

import * as S from "./styles";

const Glitch = ({ columns = "one" }: IGlitchProps) => {
  return <S.Grid columns={columns} />;
};

export default Glitch;
