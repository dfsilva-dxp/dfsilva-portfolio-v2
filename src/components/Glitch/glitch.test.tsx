import { screen } from "@testing-library/dom";
import { renderWithTheme } from "../../utils/tests/renderWithTheme";

import Glitch from ".";

describe("<Glitch/>", () => {
  it("should render a column by default", () => {
    renderWithTheme(
      <div>
        <Glitch data-testid="glitch" />
        <p>Glitch</p>
      </div>
    );

    expect(screen.getByText("Glitch").previousElementSibling).toHaveStyle({
      "grid-template-columns": "1fr"
    });
  });

  it("should render two columns when passed", () => {
    renderWithTheme(
      <div>
        <Glitch data-testid="glitch" columns="two" />
        <p>Glitch</p>
      </div>
    );

    const glitch = screen.getByText("Glitch").previousElementSibling;

    expect(glitch).toHaveStyle({
      "grid-template-columns": "repeat(2,1fr)"
    });
    expect(glitch).toHaveStyleRule("border-right", "1px solid #E1E1E6", {
      modifier: "::before"
    });
  });

  it("should render three columns when passed", () => {
    renderWithTheme(
      <div>
        <Glitch data-testid="glitch" columns="three" />
        <p>Glitch</p>
      </div>
    );

    const glitch = screen.getByText("Glitch").previousElementSibling;

    expect(glitch).toHaveStyle({
      "grid-template-columns": "repeat(3,1fr)"
    });
    expect(glitch).toHaveStyleRule("border-right", "1px solid #E1E1E6", {
      modifier: "::before"
    });
    expect(glitch).toHaveStyleRule("border-right", "1px solid #E1E1E6", {
      modifier: "::after"
    });

    expect(glitch).toMatchSnapshot();
  });
});
