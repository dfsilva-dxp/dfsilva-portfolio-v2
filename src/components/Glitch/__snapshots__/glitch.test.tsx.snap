// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Glitch/> should render three columns when passed 1`] = `
.c0 {
  position: absolute;
  height: 100%;
  border-left: 1px solid #f2f2f2;
  border-right: 1px solid #f2f2f2;
  left: 1rem;
  right: 1rem;
  z-index: 5;
  display: grid;
  grid-template-columns: repeat(3,1fr);
}

.c0::before,
.c0::after {
  content: "";
  border-right: 1px solid #f2f2f2;
}

<div
  class="c0"
/>
`;
